### ISSUE TITLE
CentOS CI's C8S job randomly times out during boot with multipath

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#22920]
**systemd version the issue has been seen with**
 > latest main

**Used distribution**
 > CentOS Stream 8

In the recent couple of days some of the C8S jobs abort during reboot caused by a time out. Closer inspection reveals that dracut stalls the boot for a while before discovering the multipath rootfs. However, the dracut's version haven't changed for a while on C8S so I wonder if it's not caused by some of the recent patches.

```
[    8.794182] multipathd[672]: --------start up--------
[    8.794182] multipathd[672]: read /etc/multipath.conf
[    8.797702] multipathd[672]: failed to increase buffer size
[    8.798386] multipathd[672]: path checkers start up
[    8.827617] kernel: device-mapper: multipath service-time: version 0.3.0 loaded
[    8.850143] dracut-initqueue[695]: Scanning devices sda2 sdb2  for LVM logical volumes vg_n35/root vg_n35/swap
[    8.855366] multipathd[672]: mpatha: load table [0 467664896 multipath 0 0 1 1 service-time 0 2 1 8:0 1 8:16 1]
[    8.913834] dracut-initqueue[733]: Volume group "vg_n35" not found
[    8.913834] dracut-initqueue[733]: Cannot process volume group vg_n35
[    9.056417] systemd[1]: Found device /dev/mapper/vg_n35-swap.
[    9.057946] systemd[1]: Starting Resume from hibernation using device /dev/mapper/vg_n35-swap...
[    9.064928] systemd-hibernate-resume[758]: Could not resume from '/dev/mapper/vg_n35-swap' (253:3).
[    9.064889] kernel: PM: Image not found (code -22)
[    9.065890] systemd[1]: systemd-hibernate-resume@dev-mapper-vg_n35\x2dswap.service: Deactivated successfully.
[    9.066164] systemd[1]: Finished Resume from hibernation using device /dev/mapper/vg_n35-swap.
[    9.066295] audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=systemd-hibernate-resume@dev-mapper-vg_n35\x2dswap comm="systemd" exe="/usr/lib/systemd/systemd" hostnam>
[    9.066383] audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=systemd-hibernate-resume@dev-mapper-vg_n35\x2dswap comm="systemd" exe="/usr/lib/systemd/systemd" hostname>
[    9.066463] systemd[1]: Reached target Preparation for Local File Systems.
[    9.066541] systemd[1]: Reached target Local File Systems.
[    9.067829] systemd[1]: Starting Create Volatile Files and Directories...
[    9.075069] systemd-tmpfiles[759]: 8;;file://n35.pufty.ci.centos.org/usr/lib/tmpfiles.d/systemd.conf^G/usr/lib/tmpfiles.d/systemd.conf8;;^G:19: Failed to resolve user 'systemd-network': No such process
[    9.075453] systemd-tmpfiles[759]: 8;;file://n35.pufty.ci.centos.org/usr/lib/tmpfiles.d/systemd.conf^G/usr/lib/tmpfiles.d/systemd.conf8;;^G:20: Failed to resolve user 'systemd-network': No such process
[    9.075530] systemd-tmpfiles[759]: 8;;file://n35.pufty.ci.centos.org/usr/lib/tmpfiles.d/systemd.conf^G/usr/lib/tmpfiles.d/systemd.conf8;;^G:21: Failed to resolve user 'systemd-network': No such process
[    9.075599] systemd-tmpfiles[759]: 8;;file://n35.pufty.ci.centos.org/usr/lib/tmpfiles.d/systemd.conf^G/usr/lib/tmpfiles.d/systemd.conf8;;^G:22: Failed to resolve user 'systemd-network': No such process
[    9.079361] systemd[1]: Finished Create Volatile Files and Directories.
[    9.079460] audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=systemd-tmpfiles-setup comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=su>
[    9.079615] systemd[1]: Reached target System Initialization.
[    9.079698] systemd[1]: Reached target Basic System.
[  147.147105] dracut-initqueue[485]: Warning: dracut-initqueue timeout - starting timeout scripts
[  147.191534] dracut-initqueue[9697]: Scanning devices sda2 sdb2  for LVM logical volumes vg_n35/root vg_n35/swap
[  147.209162] dracut-initqueue[9722]: ACTIVE '/dev/vg_n35/swap' [<2.09 GiB] inherit
[  147.209578] dracut-initqueue[9722]: inactive '/dev/vg_n35/root' [220.42 GiB] inherit
[  147.349202] systemd[1]: Found device /dev/mapper/vg_n35-root.
[  147.349584] systemd[1]: Reached target Initrd Root Device.
[  147.484290] systemd[1]: Finished dracut initqueue hook.
[  147.484439] audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=dracut-initqueue comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  147.484583] systemd[1]: Reached target Preparation for Remote File Systems.
[  147.484436] kernel: kauditd_printk_skb: 4 callbacks suppressed
[  147.484441] kernel: audit: type=1130 audit(1648722193.712:22): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=dracut-initqueue comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? add>
[  147.512466] systemd[1]: Reached target Remote File Systems.
[  147.512603] systemd[1]: dracut pre-mount hook was skipped because all trigger condition checks failed.
[  147.513915] systemd[1]: Starting File System Check on /dev/mapper/vg_n35-root...
[  147.540426] systemd-fsck[9747]: /usr/sbin/fsck.xfs: XFS file system.
[  147.542086] systemd[1]: Finished File System Check on /dev/mapper/vg_n35-root.
[  147.542242] audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=systemd-fsck-root comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  147.543576] systemd[1]: Mounting /sysroot...
```

Full debug journal from one of the boots: https://mrc0mmand.fedorapeople.org/c8s-timeout.journal

/cc @bluca @yuwata @keszybz

**Comment by mrc0mmand** [comment]
It may, theoretically, be related to https://github.com/systemd/systemd/issues/22911, as it started failing recently as well. I'll try to bisect it.

**Comment by bluca** [comment]
Nicely found! One for @keszybz I guess?

**Comment by mrc0mmand** [comment]
Okay, after trying this on a different machine I can't reproduce it with the aforementioned commit, *sigh*. I grabbed me a beer and give it another try.

**Comment by mrc0mmand** [comment]
Nevermind. It apparently doesn't reproduce on all machines in the CentOS CI pools, but on those were it does the issue goes away after reverting d6c9411072901556176ac130f2ce71a33107aa93.

**Comment by mrc0mmand** [comment]
Boot journals with `rd.debug`:

Failing run: https://mrc0mmand.fedorapeople.org/mpath-fail.journal
Passing run: https://mrc0mmand.fedorapeople.org/mpath-pass.journal

(didn't went through them thoroughly, yet, will do that in the morning)

**Comment by mrc0mmand** [comment]
One difference I can spot is this part in the failing run:

```
[   12.985220] systemd[1]: dev-dm\x2d3.device: Changed dead -> plugged
[   12.985373] systemd[1]: sys-devices-virtual-block-dm\x2d3.device: Changed dead -> plugged
[   12.985528] systemd[1]: dev-disk-by\x2duuid-e0451daa\x2dff6f\x2d4e98\x2da5ae\x2d1b08c1626282.device: Device dev-disk-by\x2duuid-e0451daa\x2dff6f\x2d4e98\x2da5ae\x2d1b08c1626282.device appeared twice with different sysfs paths /sys/devices/pci0000:00/0000:00:06.0/0000:05:00.0/ata17/host16/target16:0:0/16:0:0:0/block/sdb/sdb1 and /sys/devices/virtual/block/dm-1, ignoring the latter.
[   12.985695] systemd[1]: dev-disk-by\x2dpartuuid-bc55c485\x2d01.device: Device dev-disk-by\x2dpartuuid-bc55c485\x2d01.device appeared twice with different sysfs paths /sys/devices/pci0000:00/0000:00:06.0/0000:05:00.0/ata17/host16/target16:0:0/16:0:0:0/block/sdb/sdb1 and /sys/devices/virtual/block/dm-1, ignoring the latter.
[   12.985856] systemd[1]: dev-disk-by\x2did-wwn\x2d0xMvD1335094BB7D7_1335094BB7D7\x2dpart1.device: Changed dead -> plugged
[   12.986023] systemd[1]: dev-mapper-mpatha1.device: Changed dead -> plugged
[   12.986178] systemd[1]: dev-disk-by\x2did-scsi\x2dSMvD1335094BB7D7_1335094BB7D7\x2dpart1.device: Changed dead -> plugged
[   12.986334] systemd[1]: dev-disk-by\x2did-dm\x2dname\x2dmpatha1.device: Changed dead -> plugged
[   12.986492] systemd[1]: dev-disk-by\x2did-dm\x2duuid\x2dpart1\x2dmpath\x2dSMvD1335094BB7D7_1335094BB7D7.device: Changed dead -> plugged
[   12.986649] systemd[1]: dev-dm\x2d1.device: Changed dead -> plugged
[   12.986803] systemd[1]: sys-devices-virtual-block-dm\x2d1.device: Changed dead -> plugged
[   12.986962] systemd[1]: lvm2-pvscan@253:2.service: Failed to load configuration: No such file or directory
[   12.987171] systemd[1]: dev-disk-by\x2dpartuuid-bc55c485\x2d02.device: Device dev-disk-by\x2dpartuuid-bc55c485\x2d02.device appeared twice with different sysfs paths /sys/devices/pci0000:00/0000:00:06.0/0000:05:00.0/ata17/host16/target16:0:0/16:0:0:0/block/sdb/sdb2 and /sys/devices/virtual/block/dm-2, ignoring the latter.
[   12.987338] systemd[1]: dev-disk-by\x2did-lvm\x2dpv\x2duuid\x2dR2dCpz\x2dtO3b\x2dh227\x2d1fzf\x2dEfhe\x2dsSuf\x2doG8hhW.device: Device dev-disk-by\x2did-lvm\x2dpv\x2duuid\x2dR2dCpz\x2dtO3b\x2dh227\x2d1fzf\x2dEfhe\x2dsSuf\x2doG8hhW.device appeared twice with different sysfs paths /sys/devices/pci0000:00/0000:00:06.0/0000:05:00.0/ata17/host16/target16:0:0/16:0:0:0/block/sdb/sdb2 and /sys/devices/virtual/block/dm-2, ignoring the latter.
[   12.987506] systemd[1]: dev-disk-by\x2did-scsi\x2dSMvD1335094BB7D7_1335094BB7D7\x2dpart2.device: Changed dead -> plugged
```

which sounds relevant.

**Comment by yuwata** [comment]
cc @keszybz

**Comment by mrc0mmand** [comment]
After more testing, the issue is indeed caused by this hunk:

```diff
@@ -1936,11 +1936,15 @@ static int install_info_symlink_wants(
                         continue;
                 }
 
-                path = strjoin(config_path, "/", dst, suffix, n);
+                _cleanup_free_ char *path = strjoin(config_path, "/", dst, suffix, n);
                 if (!path)
                         return -ENOMEM;
 
-                q = create_symlink(lp, info->path, path, true, changes, n_changes);
+                _cleanup_free_ char *target = strjoin("../", info->name);
+                if (!target)
+                        return -ENOMEM;
+
+                q = create_symlink(lp, target, path, true, changes, n_changes);
                 if (r == 0)
                         r = q;
```

but I have no idea why...

**Comment by mrc0mmand** [comment]
Ugh, the plot thickens. After playing around a bit and applying this glorious patch:

```diff
diff --git a/src/shared/install.c b/src/shared/install.c
index 8e81ec41bf..3b0ada1d64 100644
--- a/src/shared/install.c
+++ b/src/shared/install.c
@@ -2002,11 +2002,17 @@ static int install_info_symlink_wants(
                 if (!path)
                         return -ENOMEM;
 
-                _cleanup_free_ char *target = strjoin("../", info->name);
-                if (!target)
-                        return -ENOMEM;
+                if (endswith(path, "multipathd.service")) {
+                        log_warning("DEBUG: not creating a relative symlink for unit %s", path);
+                        q = create_symlink(lp, info->path, path, true, changes, n_changes);
+                } else {
+                        _cleanup_free_ char *target = strjoin("../", info->name);
+                        if (!target)
+                                return -ENOMEM;
 
-                q = create_symlink(lp, target, path, true, changes, n_changes);
+                        log_warning("DEBUG: creating a relative symlink %s for unit %s", target, path);
+                        q = create_symlink(lp, target, path, true, changes, n_changes);
+                }
                 if (r == 0)
                         r = q;
```

everything starts to magically work, i.e. the machine boots fine and no "....device appeared twice..." messages in the journal.

Here's the state of the *.wants* dropins from such boot:

```
[    6.056754] systemd[1]: DEBUG: loading dropin /run/systemd/generator/initrd.target.wants/dev-mapper-vg_n11\x2droot.device for unit initrd.target
[    6.069777] systemd[1]: DEBUG: dropin /run/systemd/generator/initrd.target.wants/dev-mapper-vg_n11\x2droot.device for unit initrd.target points to ../dev-mapper-vg_n11\x2droot.device
[    6.086108] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-cmdline-ask.service for unit initrd.target
[    6.098338] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-cmdline-ask.service for unit initrd.target points to ../dracut-cmdline-ask.service
[    6.113365] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-cmdline.service for unit initrd.target
[    6.125245] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-cmdline.service for unit initrd.target points to ../dracut-cmdline.service
[    6.139561] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-initqueue.service for unit initrd.target
[    6.151613] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-initqueue.service for unit initrd.target points to ../dracut-initqueue.service
[    6.166297] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-mount.service for unit initrd.target
[    6.178009] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-mount.service for unit initrd.target points to ../dracut-mount.service
[    6.191998] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-pre-mount.service for unit initrd.target
[    6.204061] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-pre-mount.service for unit initrd.target points to ../dracut-pre-mount.service
[    6.218730] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-pre-pivot.service for unit initrd.target
[    6.230783] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-pre-pivot.service for unit initrd.target points to ../dracut-pre-pivot.service
[    6.245449] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-pre-trigger.service for unit initrd.target
[    6.257676] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-pre-trigger.service for unit initrd.target points to ../dracut-pre-trigger.service
[    6.272701] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/dracut-pre-udev.service for unit initrd.target
[    6.284665] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/dracut-pre-udev.service for unit initrd.target points to ../dracut-pre-udev.service
[    6.299161] systemd[1]: DEBUG: loading dropin /etc/systemd/system/initrd.target.wants/memstrack.service for unit initrd.target
[    6.310607] systemd[1]: DEBUG: dropin /etc/systemd/system/initrd.target.wants/memstrack.service for unit initrd.target points to ../memstrack.service
[    6.417269] systemd[1]: DEBUG: loading dropin /usr/lib/systemd/system/sysinit.target.wants/kmod-static-nodes.service for unit sysinit.target
[    6.429940] systemd[1]: DEBUG: dropin /usr/lib/systemd/system/sysinit.target.wants/kmod-static-nodes.service for unit sysinit.target points to ../kmod-static-nodes.service
[    6.445307] systemd[1]: DEBUG: loading dropin /etc/systemd/system/sysinit.target.wants/multipathd-configure.service for unit sysinit.target
[    6.457882] systemd[1]: DEBUG: dropin /etc/systemd/system/sysinit.target.wants/multipathd-configure.service for unit sysinit.target points to ../multipathd-configure.service
[    6.473423] systemd[1]: DEBUG: loading dropin /etc/systemd/system/sysinit.target.wants/multipathd.service for unit sysinit.target
[    6.485127] systemd[1]: DEBUG: dropin /etc/systemd/system/sysinit.target.wants/multipathd.service for unit sysinit.target points to /usr/lib/systemd/system/multipathd.service
[    6.500765] systemd[1]: DEBUG: loading dropin /usr/lib/systemd/system/sysinit.target.wants/plymouth-start.service for unit sysinit.target
[    6.513174] systemd[1]: DEBUG: dropin /usr/lib/systemd/system/sysinit.target.wants/plymouth-start.service for unit sysinit.target points to ../plymouth-start.service
[    6.528020] systemd[1]: DEBUG: loading dropin /usr/lib/systemd/system/sysinit.target.wants/systemd-ask-password-console.path for unit sysinit.target
[    6.541386] systemd[1]: DEBUG: dropin /usr/lib/systemd/system/sysinit.target.wants/systemd-ask-password-console.path for unit sysinit.target points to ../systemd-ask-password-console.path
[    6.558158] systemd[1]: DEBUG: loading dropin /run/systemd/generator/sysinit.target.wants/systemd-hibernate-resume@dev-mapper-vg_n11\x2dswap.service for unit sysinit.target
[    6.573607] systemd[1]: DEBUG: dropin /run/systemd/generator/sysinit.target.wants/systemd-hibernate-resume@dev-mapper-vg_n11\x2dswap.service for unit sysinit.target points to /usr/lib/systemd/system/systemd-hibernate-resume@.service
```

*w h a t*

**Comment by mrc0mmand** [comment]
As discussed on the meeting, here's the initrd: https://mrc0mmand.fedorapeople.org/centosci-multipath-initramfs-4.18.0-373.el8.x86_64.img

Also, output from one of the "reproducer" runs:

```
$ ./test.sh n4.pufty
+ HOST=n4.pufty
+ for i in '{0..19}'
+ echo 'Iteration #0'
Iteration #0
+ ssh root@n4.pufty 'systemctl reboot'
The authenticity of host 'n4.pufty (172.19.3.68)' can't be established.
ECDSA key fingerprint is SHA256:pNpn0cZERaqU1sJP3pwcuc+sd4C/xv89m0K5mF0x6Zg.
ECDSA key fingerprint is MD5:a4:a9:7d:ee:9d:39:b7:d2:71:10:66:3a:74:76:ba:a4.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'n4.pufty,172.19.3.68' (ECDSA) to the list of known hosts.
+ sleep 90
+ ssh root@n4.pufty 'systemd-analyze blame | head -n5'
4.214s dracut-initqueue.service
2.340s kdump.service
1.713s systemd-udev-settle.service
1.526s initrd-switch-root.service
1.419s systemd-random-seed.service
+ for i in '{0..19}'
+ echo 'Iteration #1'
Iteration #1
+ ssh root@n4.pufty 'systemctl reboot'
+ sleep 90
+ ssh root@n4.pufty 'systemd-analyze blame | head -n5'
4.195s dracut-initqueue.service
2.357s kdump.service
1.720s systemd-udev-settle.service
1.532s initrd-switch-root.service
1.386s systemd-random-seed.service
+ for i in '{0..19}'
+ echo 'Iteration #2'
Iteration #2
+ ssh root@n4.pufty 'systemctl reboot'
+ sleep 90
+ ssh root@n4.pufty 'systemd-analyze blame | head -n5'
ssh: connect to host n4.pufty port 22: No route to host
+ echo 'Failed on iteration #2'
Failed on iteration #2
+ exit 1

## After waiting a couple of minutes
$ ssh root@n4.pufty 'systemd-analyze blame | head -n5'
2min 21.205s dracut-initqueue.service
      2.385s kdump.service
      1.722s systemd-udev-settle.service
      1.517s initrd-switch-root.service
      1.366s systemd-random-seed.service
```

**Comment by yuwata** [comment]
Is this fixed by #23063?

**Comment by mrc0mmand** [comment]
Unfortunately no, the last fail caused by this issue is from today (Apr 18, 2022, 4:17:44 PM).
